### PR TITLE
L1T EG Phase2 Track match and EG validation code

### PR DESCRIPTION
**Pull request by cerminar**
#### PR description:

The PR addresses a minor problem in the matching of L1 tracks & EG standalone objects (see [Jack Li talk @ EG L1 P2 meeting ](https://indico.cern.ch/event/933640/contributions/3923081/attachments/2066176/3467521/EgammaMeeting_30Jun.pdf) for details).

The net effect of the fix is to increase the efficiency for central electrons at no expenses in rate (see attached plots).
The new code also allows (optionally) to use a different (more efficient) strategy for the matching using the cluster pT in the track extrapolation. The code is however deactivated for now pending FW demonstration.

The PR also integrates some validation code (ntuples) not consumed in any production workflow.

#### PR validation:

canvas(168)

canvas(169)

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TCaloTriggerNtupleBase.h [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include "L1Trigger/L1THGCalUtilities/interface/HGCalTriggerNtupleBase.h"
+
+class L1TCaloTriggerNtupleBase : public HGCalTriggerNtupleBase {
+public:
+  L1TCaloTriggerNtupleBase(const edm::ParameterSet& conf)
+      : HGCalTriggerNtupleBase(conf),
+        branch_name_prefix_(conf.getUntrackedParameter<std::string>("BranchNamePrefix", "")) {}
+  ~L1TCaloTriggerNtupleBase() override{};
+
+  char const* branch_name_w_prefix(const std::string name) const { return (branch_name_prefix_ + "_" + name).c_str(); }
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TCaloTriggerNtupleBase.h:10]: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-08-25-1100/src/L1Trigger/L1CaloTrigger/test/ntuples/L1TCaloTriggerNtupleBase.h:10:75: warning: returning address of local temporary object [-Wreturn-stack-address]

**Comment by kpedro88** [comment]
please test

File: L1Trigger/L1CaloTrigger/python/L1TCaloTriggerNtuples_cff.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+ntuple_multiclusters_hmvdr = ntuple_multiclusters.clone()
+ntuple_multiclusters_hmvdr.Prefix = cms.untracked.string('HMvDR')
+
+l1CaloTriggerNtuplizer.Ntuples = cms.VPSet(ntuple_event,
+                                           ntuple_gen,
+                                           ntuple_triggercells,
+                                           ntuple_multiclusters_hmvdr)
+
+# l1CaloTriggerNtuplizer.Ntuples.remove(ntuple_genjet)
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/python/L1TCaloTriggerNtuples_cff.py:15]: delete commented-out code (or add comment explaining why it is kept)

File: L1Trigger/L1CaloTrigger/python/L1TCaloTriggerNtuples_cff.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+l1CaloTriggerNtuplizer.Ntuples.append(ntuple_TTTracks)
+l1CaloTriggerNtuplizer.Ntuples.append(ntuple_tkEleEllEE)
+l1CaloTriggerNtuplizer.Ntuples.append(ntuple_tkEleEllEB)
+
+#
+
+l1CaloTriggerNtuples = cms.Sequence(l1CaloTriggerNtuplizer)
+
+
+# from Configuration.Eras.Modifier_phase2_hgcalV9_cff import phase2_hgcalV9
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/python/L1TCaloTriggerNtuples_cff.py:39]: delete commented-out code (or add comment explaining why it is kept)

File: L1Trigger/L1CaloTrigger/python/l1EgammaStaProducers_cff.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
+
+# from L1Trigger.L1CaloTrigger.l1EGammaCrystalsProducer_cfi import *
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/python/l1EgammaStaProducers_cff.py:3]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/BuildFile.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  <use   name="L1Trigger/L1THGCal"/>
+  <use   name="L1Trigger/L1THGCalUtilities"/>
+  <use   name="CommonTools/BaseParticlePropagator"/>
+  <use   name="MagneticField/Engine"/>
+  <use   name="MagneticField/Records"/>
+  <use   name="L1Trigger/Phase2L1ParticleFlow"/>
+  <use   name="L1Trigger/L1TTrackMatch"/>
+  <use   name="Geometry/TrackerGeometryBuilder"/>
+
+  <!-- <use   name="CommonTools/UtilAlgos"/>
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/BuildFile.xml:11]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleEgamma.cc [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  tree.Branch(branch_name_w_prefix("hwQual").c_str(), &egamma_hwQual_);
+}
+
+void L1TriggerNtupleEgamma::fill(const edm::Event& e, const edm::EventSetup& es) {
+  // retrieve towers
+  edm::Handle<l1t::EGammaBxCollection> egamma_h;
+  e.getByToken(egamma_token_, egamma_h);
+  const l1t::EGammaBxCollection& egamma_collection = *egamma_h;
+
+  // triggerTools_.eventSetup(es);
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleEgamma.cc:45]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTkElectrons.cc [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  tree.Branch(branch_name_w_prefix("tkIso").c_str(), &tkEle_tkIso_);
+}
+
+void L1TriggerNtupleTkElectrons::fill(const edm::Event& e, const edm::EventSetup& es) {
+  // retrieve towers
+  edm::Handle<l1t::TkElectronCollection> tkEle_h;
+  e.getByToken(tkEle_token_, tkEle_h);
+  const l1t::TkElectronCollection& tkEle_collection = *tkEle_h;
+
+  // triggerTools_.eventSetup(es);
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTkElectrons.cc:52]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+public:
+  L1TriggerNtupleTrackTrigger(const edm::ParameterSet& conf);
+  ~L1TriggerNtupleTrackTrigger() override{};
+  void initialize(TTree&, const edm::ParameterSet&, edm::ConsumesCollector&&) final;
+  void fill(const edm::Event& e, const edm::EventSetup& es) final;
+  typedef TTTrack<Ref_Phase2TrackerDigi_> L1TTTrackType;
+
+private:
+  void clear() final;
+  // HGCalTriggerTools triggerTools_;
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:32]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+                                          double iCharge,
+                                          double iBField);
+
+  edm::EDGetToken track_token_;
+
+  int l1track_n_;
+  std::vector<float> l1track_pt_;
+  std::vector<float> l1track_pt2stubs_;
+
+  // std::vector<float> l1track_energy_;
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:44]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+void L1TriggerNtupleTrackTrigger::fill(const edm::Event& ev, const edm::EventSetup& es) {
+  // the L1Tracks
+  edm::Handle<std::vector<L1TTTrackType>> l1TTTrackHandle;
+  ev.getByToken(track_token_, l1TTTrackHandle);
+
+  float fBz = 0;
+  if (magfield_watcher_.check(es)) {
+    edm::ESHandle<MagneticField> magfield;
+    es.get<IdealMagneticFieldRecord>().get(magfield);
+    // aField_ = &(*magfield);
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:97]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+  triggerTools_.eventSetup(es);
+
+  clear();
+  for (auto trackIter = l1TTTrackHandle->begin(); trackIter != l1TTTrackHandle->end(); ++trackIter) {
+    l1track_n_++;
+    l1track_pt_.emplace_back(trackIter->momentum().perp());
+    l1track_pt2stubs_.emplace_back(pTFrom2Stubs::pTFrom2(trackIter, tGeom));
+
+    // l1track_energy_.emplace_back(trackIter->energy());
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:114]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    l1track_eta_.emplace_back(trackIter->momentum().eta());
+    l1track_phi_.emplace_back(trackIter->momentum().phi());
+    l1track_curv_.emplace_back(trackIter->rInv());
+    l1track_chi2_.emplace_back(trackIter->chi2());
+    l1track_chi2Red_.emplace_back(trackIter->chi2Red());
+    l1track_nStubs_.emplace_back(trackIter->getStubRefs().size());
+    // FIXME: need to be configuratble?
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:121]: should this be fixed?

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    l1track_caloeta_.emplace_back(caloetaphi.first);
+    l1track_calophi_.emplace_back(caloetaphi.second);
+  }
+}
+
+void L1TriggerNtupleTrackTrigger::clear() {
+  l1track_n_ = 0;
+  l1track_pt_.clear();
+  l1track_pt2stubs_.clear();
+  // l1track_energy_.clear();
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:144]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+  l1track_chi2_.clear();
+  l1track_chi2Red_.clear();
+  l1track_nStubs_.clear();
+  l1track_z0_.clear();
+  l1track_charge_.clear();
+  l1track_caloeta_.clear();
+  l1track_calophi_.clear();
+}
+
+// #include "FastSimulation/Particle/interface/RawParticle.h"
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:157]: delete commented-out code

File: L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+                                                                     const math::XYZTLorentzVector& iVtx,
+                                                                     double iCharge,
+                                                                     double iBField) {
+  BaseParticlePropagator prop = BaseParticlePropagator(RawParticle(iMom, iVtx, iCharge), 0., 0., iBField);
+  prop.setPropagationConditions(129.0, triggerTools_.getLayerZ(1), false);
+  prop.propagate();
+  double ecalShowerDepth = reco::PFCluster::getDepthCorrection(prop.particle().momentum().E(), false, false);
+  math::XYZVector point = math::XYZVector(prop.particle().vertex()) +
+                          math::XYZTLorentzVector(prop.particle().momentum()).Vect().Unit() * ecalShowerDepth;
+  // math::XYZVector point  = particle.vertex();
```
Comment (kpedro88) [inline L1Trigger/L1CaloTrigger/test/ntuples/L1TriggerNtupleTrackTrigger.cc:169]: delete commented-out code

File: L1Trigger/L1TTrackMatch/python/L1TkEgammaObjects_cff.py [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import FWCore.ParameterSet.Config as cms
+
+from L1Trigger.L1TTrackMatch.L1TkElectronTrackProducer_cfi import *
+
+l1TkElectronTrackEllipticProducers = cms.Sequence(L1TkElectronsEllipticMatchCrystal+L1TkElectronsEllipticMatchHGC)
+l1TkElectronTrackEllipticProducersEB = cms.Sequence(L1TkElectronsEllipticMatchCrystal)
+l1TkElectronTrackEllipticProducersEE = cms.Sequence(L1TkElectronsEllipticMatchHGC)
+
+# from L1Trigger.L1TTrackMatch.L1TkEmParticleProducer_cfi import *
```
Comment (kpedro88) [inline L1Trigger/L1TTrackMatch/python/L1TkEgammaObjects_cff.py:9]: delete commented-out code

File: L1Trigger/L1TTrackMatch/python/L1TkElectronTrackProducer_cfi.py [hunk 17]
```diff
@@ -79,11 +79,11 @@
 
 
 L1TkElectronsEllipticMatchHGC = L1TkElectronsHGC.clone(
-#     L1TrackInputTag = cms.InputTag("TTTracksFromTrackletEmulation", "Level1TTTracks"),
+    # L1TrackInputTag = cms.InputTag("TTTracksFromTrackletEmulation", "Level1TTTracks")
```
Comment (kpedro88) [inline L1Trigger/L1TTrackMatch/python/L1TkElectronTrackProducer_cfi.py:82]: delete commented-out code

**Comment by cerminar** [comment]
Hello, is there any remaining issue with this?
Best,
G

**Comment by kpedro88** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade

**Comment by rekovic** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**